src/main/java/org/testng/TestRunner.java

```diff
@@ -99,16 +102,10 @@
       Maps.newHashMap();
   private final DataProviderHolder holder = new DataProviderHolder();
 
-  /**
-   * All the test methods we found, associated with their respective classes. Note that these test
-   * methods might belong to different classes. We pick which ones to run at runtime.
-   */
-  private ITestNGMethod[] m_allTestMethods = new ITestNGMethod[0];
-
-  // Information about this test run
-
   private Date m_startDate = new Date();
   private Date m_endDate = null;
+  private final IContainer<ITestNGMethod> testMethodsContainer =
+      new TestMethodContainer(this::computeAndGetAllTestMethods);
```

Comment:
Awesome, you did it :D

Reply:
yeah, the hint that you shared helped :) and once I added it, it became more easy to toggle between cache and on-demand evaluation.